src/_pytest/main.py

```diff
@@ -295,7 +296,9 @@ def _in_venv(path):
     return any([fname.basename in activates for fname in bindir.listdir()])
 
 
-def pytest_ignore_collect(path, config):
+def pytest_ignore_collect(
+    path: py.path.local, config: Config
+) -> "Optional[Literal[True]]":
```

Comment:
@bluetech I assume using Literal[True] here makes sense now?  (although not in the hookspec itself)
(I've quickly checked your other typing PRs, but it was not included there, is it?)

Reply:
General question: is the hook *implementation* allowed to declare a return type that is different, but still a subset, of the return type declared by the *hookspec*?

Example:

```python
@hookspec
def pytest_ignore_collect(
    path: py.path.local, config: Config
) -> "Optional[bool]":
    ...

@hookimpl
def pytest_ignore_collect(
    path: py.path.local, config: Config
) -> "Optional[Literal[True]]":
    ...
```

Reply:
@nicoddemus 
Yes, this is allowed in general:
```python
from typing import Literal


class A:
    def f(self) -> bool:
        pass


class B(A):
    def f(self) -> Literal[True]:
        return True
```

(keep in mind though that mypy currently does not know that hookimpls are subsets of hookspecs (as with subclasses))

Reply:
@blueyed 

> @bluetech I assume using Literal[True] here makes sense now? (although not in the hookspec itself)

There are arguments for and against.

For: the type provides a quick indication that this impl either only returns `None` or `True`. Can gain useful semantic information just from the type, namely that it doesn't return `False`.

Against: If sometime the circumstances change, and the function wants to start returning `False`, that's perfectly legit logically, but the type needs to change. This suggests the type is over restrictive.

Against: Might confuse the person who wants to make the change, if they don't realize they are *allowed* to extend the type. Might confuse copy/pasters. (Note: the argument is weakened by the fact hookimpls are already allowed to omit arguments they don't use, so might not match the hookspec already).

My inclination is to use the more general type and match the hookspec, but I can see both ways, and am probably not too consistent myself. So I'm fine with whatever you decide.

Reply:
@bluetech 
Thanks for elaborating on it again.  I think the benefit of gaining useful semantic information just from the type is useful here in general.

Will wait for any other feedback, and otherwise merge it as is.

Reply:
> Yes, this is allowed in general

Yes thanks, that's exactly what I meant to ask (not about hooks/impls, but in general).

> Will wait for any other feedback, and otherwise merge it as is.

I'm fine with what you guys decide, and the current form. 👍